hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/impl/RebuiltState.java

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.state.impl;
+
+import com.hedera.node.app.spi.state.ReadableState;
+import com.hedera.node.app.state.merkle.HederaStateImpl;
+import com.swirlds.common.merkle.MerkleNode;
+import com.swirlds.fchashmap.FCHashMap;
+import java.util.Map;
+import java.util.Objects;
+import javax.annotation.Nonnull;
+
+/**
+ * An implementation of {@link ReadableState} backed by a {@link FCHashMap} for aliases, that needs
+ * to be rebuilt from the {@link HederaStateImpl}.
+ *
+ * @param <K> The type of key for the state
+ * @param <V> The type of value for the state
+ */
+public final class RebuiltState<K, V> extends MutableStateBase<K, V> {
+    private Map<K, V> aliases;
+
+    // TODO This is broken, it also needs a reference to the state with account Ids,
```

Comment:
I don't understand this comment...imo aliases are just the (lone) example of a `RebuiltState`.

Reply:
Thanks for pointing this out. Actually, I need to think more about the rebuilt data structure case. I think we need to get rid of all rebuilt data structures, they shouldn't be a tool in our toolbox. They won't scale (in terms of memory use or startup time) as we add more and more state to the system. For example in this case I'd have a virtual map of alias -> account Id, which means in our business logic a lookup based on alias resolves to two lookups on disk. But this eliminates the scaling problem (for memory and startup) and creates a constant-time overhead for a large number of aliased accounts (and it looks like the future may include a large number of accounts having aliases).

But, in the short term at least, we have to deal with rebuilt data structures. How do we handle those? Do we make fast-copies of them each time we get a new round?